app/src/main/java/com/davidtomas/taskyapp/features/agenda/data/sync/repository/SyncRepository.kt

```diff
@@ -0,0 +1,110 @@
+    private val taskService: TaskService,
+    private val reminderService: ReminderService,
+    private val eventLocalSource: EventLocalSource,
+    private val reminderLocalSource: ReminderLocalSource,
+    private val taskLocalSource: TaskLocalSource,
+    private val agendaRepository: AgendaRepository,
+) {
+
+    suspend operator fun invoke(): Result<Unit, DataError> {
+        val deleteAgendaItems = deleteAgendaItems()
+        if (deleteAgendaItems is Result.Error) return deleteAgendaItems
+        val updateTasksResult = updateTasks()
+        if (updateTasksResult is Result.Error) return updateTasksResult
+        val updateRemindersResult = updateReminders()
+        if (updateRemindersResult is Result.Error) return updateRemindersResult
+        clearTaskyTables()
+        val fetchAgendaResult = agendaRepository.fetchAgenda()
+        if (fetchAgendaResult is Result.Error) return fetchAgendaResult
+        return Result.Success(Unit)
+    }
+
+    private suspend fun clearTaskyTables() {
+        eventLocalSource.clearRealTables()
+    }
+
+    private suspend fun deleteAgendaItems(): Result<Unit, DataError.Network> {
+        val deletedEventsIds = mutableListOf<String>()
+        val deletedTasksIds = mutableListOf<String>()
+        val deletedRemindersIds = mutableListOf<String>()
+        return coroutineScope {
+            val getDeletedEvents = launch {
+                deletedEventsIds.addAll(eventLocalSource.getUnSyncedDeletedEvents())
+            }
+            val getDeletedTasks = launch {
+                deletedTasksIds.addAll(taskLocalSource.getUnSyncedDeletedTasks())
+            }
+            val getDeletedReminders = launch {
+                deletedRemindersIds.addAll(reminderLocalSource.getUnSyncedDeletedReminder())
+            }
+
+            getDeletedEvents.join()
+            getDeletedTasks.join()
+            getDeletedReminders.join()
+            Log.d("WorkerStatus", "$deletedEventsIds $deletedRemindersIds $deletedTasksIds")
+            syncService.syncAgendaItems(
+                deletedEventsIds = deletedEventsIds,
+                deletedTasksIds = deletedTasksIds,
+                deletedRemindersIds = deletedRemindersIds
+            )
+        }
+    }
+
+    private suspend fun updateTasks(): Result<Unit, DataError.Network> {
+        val updatedUnSyncedTasks = taskLocalSource.getUnSyncedUpdatedTasks()
+        Log.d("WorkerStatus", "$updatedUnSyncedTasks")
+        if (updatedUnSyncedTasks.isNotEmpty()) {
+            updatedUnSyncedTasks.map {
+                val result = taskService.updateTask(it)
+                if (result is Result.Error) return result
+            }
+        }
+        val createdUnSyncedTasks = taskLocalSource.getUnSyncedCreatedTasks()
+        Log.d("WorkerStatus", "$createdUnSyncedTasks")
+        if (createdUnSyncedTasks.isNotEmpty()) {
+            createdUnSyncedTasks.map {
+                val result = taskService.createTask(it)
```

Comment:
Here, you make every single API request sequentially. If one takes a second, this will therefore take 100s for all tasks if there are 100 pending syncs 😄 
Better use parallelism